### PR TITLE
stream info: add upstream connection id

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: stream info - add upstream connection id
Additional Description: Envoy Mobile is interested in surfacing some information from Stream Info up to the client for analysis purposes. It is useful to understand what streams were multiplexed over the same upstream connection and Envoy's internal connection id can be used to tie this up.
Risk Level:
Testing: pending

Signed-off-by: Jose Nino <jnino@lyft.com>

### CODE REVIEW

**Comment by junr03** [comment]
@alyssawilk, this came out of a discussion our team had about being able to correlate stream behavior grouped by underlying upstream connection. 

I wanted to put a draft up to discuss were and when to get this data. It looks to me from the existing behavior in the code here: https://github.com/envoyproxy/envoy/pull/17512/commits/d5b6e79ec851c7def100879468248069a9e52c7f#diff-b61a05a18225decb27f5620aefda1646c259e314547a7d985c369f62a684c12bR420

that the upstream connection populates its "downstream" side because the codec client (and the underlying connection) could be either downstream or upstream. So my first question to your is if this PR should support both upstream and downstream connection IDs, and fill only downstream when only a single direction makes sense.

Ultimately all I want is for the stream info to have access to the upstream connection id, and this seemed like the cleanest way to do so. However, I am not tied to this being the best way to obtain this information.

cc @goaway

File: envoy/stream_info/stream_info.h
Comment (junr03) [inline envoy/stream_info/stream_info.h:245]: pending docstring, will update once we settle on API

**Review by alyssawilk — COMMENTED** [review]
Yeah, this sounds reasonable to me.
@mattklein123 @RyanTheOptimist for any other thoughts

File: source/common/stream_info/stream_info_impl.h
Comment (alyssawilk) [inline source/common/stream_info/stream_info_impl.h:71]: add to dumpState?

File: source/common/router/upstream_request.cc
Comment (junr03) [reply source/common/router/upstream_request.cc:423]: Really good point! Thanks for pointing out.

I updated. Let me know if you think this is an appropriate place. Related I had to expose stream info for upstream request. Not sure if it was intentional that we did not have a public API for that.

**Comment by junr03** [comment]
@alyssawilk updated!

**Review by mattklein123 — APPROVED** [review]
Thanks LGTM at a high level. Will defer to @alyssawilk for further review.

File: source/common/router/upstream_request.cc
Comment (soulxu) [inline source/common/router/upstream_request.cc:425]: This only handles the http, should we handle the tcp case? I know the SSL connection info will be populated for tcpproxy also. https://github.com/envoyproxy/envoy/blob/32d40af143a0795a4fd45a5c0c4b59850e378543/source/common/tcp_proxy/tcp_proxy.cc#L531

File: source/common/network/connection_impl.cc
Comment (soulxu) [inline source/common/network/connection_impl.cc:828]: I'm not sure you need this.

Since we have two streamInfo for two sides. One for downstream, another one for upstream. We copy the upstream streamInfo 's id and SSL connection info into the downstream's streamInfo.  This is done into UpstreamRequest

So here, you set the upstream's connection id into the upstream's streamInfo, it seems needless.

**Review by soulxu — CHANGES_REQUESTED** [review]

File: source/common/router/upstream_request.cc
Comment (alyssawilk) [reply source/common/router/upstream_request.cc:425]: It's a good point - I was going to ask the same thing but it occurred to me that we might also want this for thrift and mysql and I think folks can add it as they want it.  the API says "not available" rather than "if there is no upstream connection" so I think it's simply not available yet for TCP

File: source/common/router/upstream_request.cc
Comment (alyssawilk) [reply source/common/router/upstream_request.cc:423]: yeah AFIK we were just adding APIs as needed. The only thing I think you should do, as we're in an area where upstream and downstream are intersecting, is comment or rename to make it clear which stream info is being accessed :-)

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo one comment request :-)

File: source/common/router/upstream_request.cc
Comment (junr03) [reply source/common/router/upstream_request.cc:425]: +1 to adding exposure as desired.

Comment (junr03) [reply source/common/router/upstream_request.cc:425]: +1 to adding exposure as desired.

File: source/common/network/connection_impl.cc
Comment (junr03) [reply source/common/network/connection_impl.cc:828]: @soulxu I am not sure I understand your comment. The value that gets assigned in onPoolReady in UpstreamRequest comes from the value set here. So we definitely need to set the value in the ConnectionImpl instance variable.

However, I do hear your comment about the confusion with upstream/downstream in the instance of stream info. In other places in the codebase stream info does have a defined upstream/downstream directionality. Given I am implementing this in the ClientConnection class I thought upstream was appropriate. I am going to add some more clarifying comment per @alyssawilk's [comment](https://github.com/envoyproxy/envoy/pull/17512#discussion_r681901853) above

File: source/common/router/upstream_request.cc
Comment (junr03) [reply source/common/router/upstream_request.cc:423]: @alyssawilk I was actually confused when I first read the ssl info lines above, and thought there was a bug. In my mind because we are receiving information about a connection that acts as a client connection the term Upstream made sense 😅. I added a comment that is hopefully clarifying. Lmk what you think

**Review by alyssawilk — APPROVED** [review]

**Comment by junr03** [comment]
@alyssawilk if you could re-stamp that would be great. I realized that there was more involved in cleaning up the connectionID function that used to exist. I'll clean it up in a subsequent PR, and rolled back the initial deletion.

File: source/common/network/connection_impl.cc
Comment (soulxu) [reply source/common/network/connection_impl.cc:828]: > @soulxu I am not sure I understand your comment. The value that gets assigned in onPoolReady in UpstreamRequest comes from the value set here. So we definitely need to set the value in the ConnectionImpl instance variable.
> 

Sorry, I just found you set the upstreamId of upstream's StreamInfo to the downstreamId of downstream's StreamInfo. That is opposite with the SSL connection info which above your change:

https://github.com/envoyproxy/envoy/blob/a8a2eb9fecb45e0e3e8a69f8176d801faed33c1c/source/common/router/upstream_request.cc#L420-L421

you can see the upstream SSL connection info of downstream's streamInfo is coming from the downstream SSL connection of upstream's streamInfo.

Yes, as @alyssawilk said, we interleave the upstream and downstream here. But this will be more confusing, since we have a different pattern between SSL connection info and connectionId.

File: source/common/router/upstream_request.cc
Comment (soulxu) [reply source/common/router/upstream_request.cc:425]: > It's a good point - I was going to ask the same thing but it occurred to me that we might also want this for thrift and mysql and I think folks can add it as they want it. the API says "not available" rather than "if there is no upstream connection" so I think it's simply not available yet for TCP

got it, thanks, if the API says "not available", then we are cool here. I thought only the last filter is responsible for populating those stream info.(the tcp proxy and HCM). I thought MySQL and thrift also need the tcp proxy filter. But I may not understand correctly here.

File: source/common/network/connection_impl.cc
Comment (soulxu) [reply source/common/network/connection_impl.cc:828]: Actually, the connectionID already set at https://github.com/envoyproxy/envoy/blob/b29d6543e7568a8a3e772c7909a1daa182acc670/source/common/network/connection_impl.cc#L101

So what you need to do in the onPoolReady in upstreamRequest is:
`stream_info_.setUpstreamConnectionId(info.downstreamAddressProvider().connectionID());`

Comment (alyssawilk) [reply source/common/network/connection_impl.cc:828]: actually this looks like it'd be a better option.  sorry to not merge this but I think it's worth the fix.
cc @goaway in case someone on the mobile side was waiting on this while Jose is out.
/wait

Comment (junr03) [reply source/common/network/connection_impl.cc:828]: > we interleave the upstream and downstream here. But this will be more confusing, since we have a different pattern between SSL connection info and connectionId.

@soulxu yep, I agree. I chose to use upstream here because of two reasons:
1. It made more sense to me, as really the connection is to an upstream entity.
2. I did not want to add downstreamConnectionId as well.

However, I agree with you that that creates confusion, and thanks to your suggestion we can just skip that :). Thanks a lot for taking the time to review and point to a better solution. I appreciate it!

I have updated to use the address provider.

**Comment by junr03** [comment]
@alyssawilk @soulxu I have updated with @soulxu's recommendation. We should be good now

**Comment by soulxu** [comment]
thanks for update! LGTM

**Review by snowp — APPROVED** [review]
Thanks!

**Pull request merged**
